pkg/apiserver/api_installer.go

```diff
@@ -110,62 +111,55 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
-	if a.group.OptionsExternalVersion != nil {
-		optionsExternalVersion = *a.group.OptionsExternalVersion
-	}
-
-	var resource, subresource string
-	switch parts := strings.Split(path, "/"); len(parts) {
-	case 2:
-		resource, subresource = parts[0], parts[1]
-	case 1:
-		resource = parts[0]
-	default:
-		// TODO: support deeper paths
-		return nil, fmt.Errorf("api_installer allows only one or two segment paths (resource or resource/subresource)")
+// validFullyQualifiedKind validates and returns the group version kind registered for the given
+// storage object. If the storage object is a subresource, it just return's the object's group
+// version kind without validating it's group against the group we are installing the REST
+// handlers for here.
+func (a *APIInstaller) validFullyQualifiedKind(path string, storage rest.Storage, hasSubresource bool) (unversioned.GroupVersionKind, error) {
+	object := storage.New()
+	if hasSubresource {
+		return a.group.Typer.ObjectKind(object)
 	}
-	hasSubresource := len(subresource) > 0
 
-	object := storage.New()
 	fqKinds, err := a.group.Typer.ObjectKinds(object)
 	if err != nil {
-		return nil, err
+		return unversioned.GroupVersionKind{}, err
 	}
+
 	// a given go type can have multiple potential fully qualified kinds.  Find the one that corresponds with the group
 	// we're trying to register here
 	fqKindToRegister := unversioned.GroupVersionKind{}
 	for _, fqKind := range fqKinds {
 		if fqKind.Group == a.group.GroupVersion.Group {
-			fqKindToRegister = fqKind
+			fqKindToRegister = a.group.GroupVersion.WithKind(fqKind.Kind)
 			break
 		}
 
 		// TODO This keeps it doing what it was doing before, but it doesn't feel right.
 		if fqKind.Group == extensions.GroupName && fqKind.Kind == "ThirdPartyResourceData" {
-			fqKindToRegister = fqKind
-			fqKindToRegister.Group = a.group.GroupVersion.Group
-			fqKindToRegister.Version = a.group.GroupVersion.Version
+			fqKindToRegister = a.group.GroupVersion.WithKind(fqKind.Kind)
 		}
 	}
-
-	kind := fqKindToRegister.Kind
-
 	if fqKindToRegister.IsEmpty() {
-		return nil, fmt.Errorf("unable to locate fully qualified kind for %v: found %v when registering for %v", reflect.TypeOf(object), fqKinds, a.group.GroupVersion)
+		return unversioned.GroupVersionKind{}, fmt.Errorf("unable to locate fully qualified kind for %v: found %v when registering for %v", reflect.TypeOf(object), fqKinds, a.group.GroupVersion)
 	}
+	return fqKindToRegister, nil
+}
 
-	versionedPtr, err := a.group.Creater.New(a.group.GroupVersion.WithKind(kind))
-	if err != nil {
-		return nil, err
+// restMapping returns the RESTMapping object for the given resource or the subresource based on
+// the supplied subresource overrides.
+func (a *APIInstaller) restMapping(path, resource string, hasSubresource bool, fqKindToRegister unversioned.GroupVersionKind) (*meta.RESTMapping, error) {
+	mapper := a.group.Mapper
+
+	if hasSubresource {
+		groupMeta, err := registered.Group(fqKindToRegister.Group)
```

Comment:
Reading this function, I cannot tell if Job's scale subresource will end up exposed at `apis/batch/v1/job/scale` or `apis/autoscaling/v1/job/scale`. The former is correct. Can you add comments until it is clear?